engine + structure ready to reviewed

NorthBlue333 opened the pull request.

P0ulpy committed:
engine + structure ready to reviewed

P0ulpy committed:
Clean editor

NorthBlue333 commented on .idea/discord.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Potentiellement à supprimer

NorthBlue333 commented on Engine/Engine/AssetLoader/AssetLoader.hpp:
```diff
@@ -0,0 +1,55 @@
+    class AssetLoader
+    {
+    public:
+        static TAsset* StaticLoadAsset(const std::string& path);
```
Pourquoi préfixer par Static ici ? Une raison en particulier ?

P0ulpy replied:
Au départ je voulais avoir plusieurs autres méthodes de LoadAsset (static/dynamic/lazy) mais au final je pense que le système vas s'articuler autrement donc le préfixe Static peut être enlevé

NorthBlue333 commented:
Globalement, utiliser `#pragma once` au lieu des `#ifndef`
C'est facile à personnaliser sur la suite JetBrains, https://www.jetbrains.com/help/clion/header-guards.html

NorthBlue333 commented on Engine/Engine/Core/Application/EngineApplication.hpp:
```diff
@@ -0,0 +1,39 @@
+        Timestep m_LastFrameTime = 0;
+
+        // TODO : use a custom stack where we can iterate throw
+        std::vector<ApplicationLayer*> m_layers { nullptr };
```
Mériterait peut-être d'avoir une taille fixe plutôt qu'un vecteur : il me semble pas qu'on est besoin d'une taille changeante au runtime ?

P0ulpy replied:
Je pense aussi, dans l'idée du moteur je voulais laisser la possibilité a l'utilisateur la possibilité de rajouter des layers à volonté, notamment pour avoir un `InternalLayer` (target Engine), `GameLayer (target Game),` `EditorLayer` (target Editor). le souci c'est qu'ils seraient définis dans des targets différentes (pour GameLayer il y a un mini-exemple avec `PathfinderLayer` dans les commentaires de `Game/src/main.cpp`).
Je ne suis pas certain de savoir comment m'y prendre pour avoir l'information du nombre de layers qui sont nécessaires pour pré-alloué la bonne quantité de mémoire

j'ai fait le choix d'utiliser des layers pour rendre possible ce genre d'implémentations, mais je ne suis pas certain que ce soit la meilleure solution

NorthBlue333 replied:
Je pense que c'est possible de le rendre compile time, c'est à dire un argument `int/size_t` de template

NorthBlue333 commented on Engine/Engine/Core/Application/EngineApplication.hpp:
```diff
@@ -0,0 +1,39 @@
+
+        // TODO : use a custom stack where we can iterate throw
+        std::vector<ApplicationLayer*> m_layers { nullptr };
+        ScenesLayer m_scenesLayer { m_window };
```
Je ne vois pas trop l'intérêt de le stocker 2 fois ce layer?

P0ulpy replied:
`m_layers` sert à itéré sur tous les layers mais `m_scenesLayer` c'est l'instance du ScenesLayer qui est présent dans tout les cas (d'ailleurs le { nullptr } de m_layers est une erreur je l'ai enlevé)
Mais je pense bouger l'instance du ScenesLayer dans une autre class qui gérerait l'instanciation / des-instanciation des scènes

NorthBlue333 commented on Engine/Engine/Core/Application/ScenesLayer.hpp:
```diff
@@ -0,0 +1,31 @@
+
+    private:
+        std::unique_ptr<Scene> m_activeScene;
+        sf::RenderTarget& m_renderTarget;
```
Pourquoi une ref et pas un pointeur ?

P0ulpy replied:
Aucune idée, je l'enlève, surtout que si m_renderTarget est invalide il faut que le programe plante, et si je ne dit pas de bêtises les Ref ne peuvent pas être invalides et donc ça vas pointer sur un bout de mémoire qui n'est pas a l'adresse 0

NorthBlue333 commented on Engine/Engine/Core/Components/SpriteRenderer.hpp:
```diff
@@ -0,0 +1,54 @@
+    public:
+        DECLARE_RTTI(SpriteRenderer, Engine::Component)
+
+        sf::Sprite RenderSprite;
```
Je crois que du coup on devais pas utiliser les Sprite SFML à cause de leur lifetime. A voir optionnellement

NorthBlue333 commented on Engine/Engine/Core/Components/SpriteRenderer.hpp:
```diff
@@ -0,0 +1,54 @@
+
+        void OnUpdate(const float& deltaTime)
+        {
+            if(transform)
```
Dans ce genre de cas, préférer les early return pour la lisibilité
```c++
if(!transform) {return;}
RenderSprite.setPosition(transform->Pos);
RenderSprite.setRotation(transform->Angle);
RenderSprite.setScale(transform->Scale);
```

NorthBlue333 commented on Engine/Engine/Core/Components/Transform.hpp:
```diff
@@ -0,0 +1,26 @@
+    public:
+        DECLARE_RTTI(Transform, Engine::Component)
+
+        sf::Vector2<float> Pos { .0f, .0f };
```
Définir un type `using VectorType = sf::Vector2<float>` c'est plus lisible et pratique

NorthBlue333 commented on Engine/Engine/Core/Components/Transform.hpp:
```diff
@@ -0,0 +1,26 @@
+
+        sf::Vector2<float> Pos { .0f, .0f };
+        sf::Vector2<float> Scale { 1.f, 1.f };
+        float Angle = .0f;
```
Ok pour la simplicité mais je ne sais pas si ça correspond à tous les cas d'usage

NorthBlue333 commented on Engine/Engine/Core/Inputs/WindowEvents.cpp:
```diff
@@ -0,0 +1,49 @@
+        sf::Event event {};
+        while(renderWindow.pollEvent(event))
+        {
+#ifdef IMGUI_SUPPORT
```
?

NorthBlue333 commented on Engine/Engine/Core/Inputs/WindowEvents.cpp:
```diff
@@ -0,0 +1,49 @@
+
+#include "WindowEvents.hpp"
+
+namespace Engine
```
j'imagine que ça ça va bouger avec le Signal
Donc pas de comm particulier

NorthBlue333 commented on Engine/Engine/Core/Logger/ConsoleColor.hpp:
```diff
@@ -0,0 +1,119 @@
+
+namespace ConsoleColor
+{
+    inline std::ostream& blue(std::ostream &s)
```
Je trouve qu'il y a beaucoup de duplication de code alors qu'il suffirait de faire genre : 
```c++
inline std::ostream& colored(std::ostream &s, attributes) {
        HANDLE hStdout = GetStdHandle(STD_OUTPUT_HANDLE);
        SetConsoleTextAttribute(hStdout, attributes|FOREGROUND_INTENSITY);
}
```

Et d'appeler colored dans les autres méthodes

NorthBlue333 commented on Engine/Engine/Core/Logger/ConsoleColor.hpp:
```diff
@@ -0,0 +1,119 @@
+
+#endif
+
+#ifdef LINUX
```
Le ifdef Linux je le ferais directement dans la méthode Colored par exemple, pour éviter de tout redéfinir
Mais c'est discutable

NorthBlue333 commented on Engine/Engine/Core/Logger/ConsoleColor.hpp:
```diff
@@ -0,0 +1,119 @@
+{
+    inline std::ostream& blue(std::ostream &s)
+    {
+        return s;
```
Pour les couleurs il faut print ça normalement
```
\e[1;31m This is red text \e[0m
```

`1;31` c'est la couleur pour rouge, `0` c'est le reset

return 
https://en.wikipedia.org/wiki/ANSI_escape_code#Colors (les 8 premières sont les standards)

NorthBlue333 reviewed (commented):
Première partie

NorthBlue333 commented on Engine/Engine/Core/Logger/Logger.cpp:
```diff
@@ -0,0 +1,122 @@
+
+#include "ConsoleColor.hpp"
+
+std::string CurrentDateTimeToString()
```
Potentiellement déplacer dans un fichier `utils` ou quelque chose comme ça

NorthBlue333 commented on Engine/Engine/Core/Logger/Logger.cpp:
```diff
@@ -0,0 +1,122 @@
+
+namespace Engine
+{
+    std::mutex Logger::_mutex;
```
C'est pour l'instanciation ça ?

P0ulpy replied:
Tout a fait

NorthBlue333 commented on Engine/Engine/Core/Logger/Logger.cpp:
```diff
@@ -0,0 +1,122 @@
+namespace Engine
+{
+    std::mutex Logger::_mutex;
+    std::list<LogEntry> Logger::messages;
```
Mmh stocker les messages ça peut rapidement faire une grande liste, pas sûre que ce soit utile ?

P0ulpy replied:
Je pense aussi go virer la std::list

NorthBlue333 commented on Engine/Engine/Core/RTTI/RTTI.hpp:
```diff
@@ -0,0 +1,49 @@
+    class RTTI
+    {
+    public:
+        using Typename = std::string;
```
C'est presque un keyword C++ "Typename" donc pas top top mais ok

P0ulpy replied:
je trouve aussi, j'ai mis a jour ça dans ma dernière comit du coup👌

NorthBlue333 commented:
Globalement également, je constate un manque de cohérence de nomenclature (notamment sur la casse des noms de fonctions/variables)

NorthBlue333 commented on Engine/Engine/Core/Time.cpp:
```diff
@@ -0,0 +1,14 @@
+    Timestep Time::GetTime()
+    {
+        // TODO : Implement Me
+        return 0;
```
C'était pour quoi ça ? Le delta time ?

P0ulpy replied:
Oui mais je pense virer la class Time et tout gérer dans EngineApplication

NorthBlue333 commented on Engine/Engine/Core/UUID.cpp:
```diff
@@ -0,0 +1,25 @@
+// Created by Flo on 26/11/2022.
+//
+
+#include "UUID.hpp"
```
L'uuid c'est censé être un vrai `uuid`, ici c'est plus un id juste

NorthBlue333 commented on Engine/Engine/ECS/EntitiesRegistry/ComponentSystem.hpp:
```diff
@@ -0,0 +1,276 @@
+        template <class T>
+        struct Entry
+        {
+            union EntryUnion
```
Ca peut être intéressant d'utiliser un variant du coup ?

P0ulpy replied:
Oui tout a fait, je suis en train de refaire cette partie j'utiliserais un Variant a la place

NorthBlue333 commented on Engine/Engine/ECS/EntitiesRegistry/ComponentSystem.hpp:
```diff
@@ -0,0 +1,276 @@
+//
```
Je relis pas trop trop comme ça va être modifié, je repasserai sur la PR en question

NorthBlue333 commented on Engine/Engine/ECS/Entity/Entity.hpp:
```diff
@@ -0,0 +1,56 @@
+
+    public:
+        Entity(EntityHandle handle, Scene* scene);
+        Entity(const Entity& other) = default;
```
C'est pour la copie ça ?

P0ulpy replied:
Oui mais je sais pas vraiment pourquoi j'ai implémenter ça comme ça

NorthBlue333 commented:
L'asset krab me semble inutile du coup

NorthBlue333 commented on Game/src/main.cpp:
```diff
@@ -0,0 +1,124 @@
+/*
```
Toute la partie commentée me semble à supprimer du coup

P0ulpy replied:
yes je la laisse juste pour avoir un exemple d’implémentation d'application avec l'Engine

P0ulpy committed:
ECS Compilation Fixed

Cette comit contient principalement des modifications au niveau des #includes et des modifications mineurs

Le module Engine compile maintenant sans problème

Certaines des remarques de la précédente PR ont été prise en compte

P0ulpy committed:
Little clean

P0ulpy committed:
Multiple windows fix & EngineApplication destruction fix

P0ulpy committed:
Webhook test

P0ulpy committed:
Mini fix

P0ulpy committed:
Some ECS in template fix

P0ulpy committed:
Bringed back the .AddComponent to Entity

P0ulpy committed:
Update/Render ComponentSystems Dispatch

P0ulpy committed:
Added dt and fixed EngineApp destruction error
